File: projects/dare-ni-mo-narenai-watashi-dakara-v3/project/render/build_lexical_proposal.py
import hashlib,json
from pathlib import Path

ROOT=Path(r"C:\project\musicjlpt\projects\dare-ni-mo-narenai-watashi-dakara-v3")
frames_path=ROOT/'project'/'frames.json'
frames=json.loads(frames_path.read_text(encoding='utf-8'))['frames']
sha=hashlib.sha256(frames_path.read_bytes()).hexdigest()

# token, reading, Hepburn romaji, short POS.  Boundaries intentionally follow
# learnable lexical/fixed-expression units while concatenating back to the lyric.
D={
'l006':[('誰か','だれか','dareka','疑问代词'),('に','に','ni','助词'),('染まれない','そまれない','somarenai','动词（否定形）'),('誰にも','だれにも','darenimo','疑问代词＋助词'),('なれない','なれない','narenai','动词（否定形）')],
'l007':[('夢','ゆめ','yume','名词'),('だ','だ','da','助动词'),('と','と','to','助词'),('思って','おもって','omotte','动词（て形）'),('笑ってみせた','わらってみせた','waratte miseta','固定表达')],
'l008':[('忘れた','わすれた','wasureta','动词（た形）'),('弱さ','よわさ','yowasa','名词'),('悴んでいく','かじかんでいく','kajikande iku','动词（ていく形）')],
'l009':[('寂しい','さびしい','sabishii','形容词'),('なんて','なんて','nante','助词'),('言えるわけもない','いえるわけもない','ieru wake mo nai','固定表达')],
'l010':[('さりげない','さりげない','sarigenai','形容词'),('会話','かいわ','kaiwa','名词'),('すら','すら','sura','助词'),('傷つく','きずつく','kizutsuku','动词'),('心','こころ','kokoro','名词'),('は','は','wa','助词')],
'l011':[('誰一人','だれひとり','dare hitori','副词性表达'),('誰一人','だれひとり','dare hitori','副词性表达'),('幸せにしない','しあわせにしない','shiawase ni shinai','固定表达')],
'l012':[('ねぇ','ねぇ','nee','感叹词'),('それでも','それでも','soredemo','连词'),('まだ','まだ','mada','副词'),('期待','きたい','kitai','名词'),('膨らむんだ','ふくらむんだ','fukuramunda','固定表达')],
'l013':[('すれ違った','すれちがった','surechigatta','动词（た形）'),('嘘','うそ','uso','名词'),('なんて','なんて','nante','助词'),('全部','ぜんぶ','zenbu','名词'),('捨てて','すてて','sutete','动词（て形）')],
'l014':[('大逸れて','おおそれて','oosorete','动词（て形）')],
'l015':[('ないよって','ないよって','nai yo tte','固定表达'),('ただ','ただ','tada','副词'),('聞きたくて','ききたくて','kikitakute','动词（たい形）')],
'l016':[('夢ばかりで','ゆめばかりで','yume bakari de','固定表达'),('幼くたって','おさなくたって','osanakutatte','固定表达')],
'l017':[('今日も','きょうも','kyou mo','名词＋助词'),('変われない','かわれない','kawarenai','动词（否定形）')],
'l018':[('ままだって','ままだって','mama datte','固定表达'),('いいんだ','いいんだ','iinda','固定表达')],
'l019':[('君','きみ','kimi','名词'),('と','と','to','助词'),('歌っていたい','うたっていたい','utatte itai','动词（ていたい形）')],
'l020':[('何度だって','なんどだって','nando datte','副词性表达')],
'l021':[('感情','かんじょう','kanjou','名词'),('ばっか','ばっか','bakka','助词'),('揺れ動いて','ゆれうごいて','yureugoite','动词（て形）')],
'l022':[('嫌われてきた','きらわれてきた','kirawarete kita','动词（てきた形）'),('けど','けど','kedo','接续助词')],
'l023':[('このままで','このままで','kono mama de','固定表达'),('歩いて行くんだ','あるいていくんだ','aruite ikunda','固定表达')],
'l024':[('誰にも','だれにも','darenimo','疑问代词＋助词'),('なれない','なれない','narenai','动词（否定形）'),('私','わたし','watashi','名词'),('胸張って','むねはって','mune hatte','固定表达'),('さ','さ','sa','终助词')],
'l025':[('さよなら','さよなら','sayonara','感叹词'),('さよなら','さよなら','sayonara','感叹词'),('要らない','いらない','iranai','形容词'),('全部','ぜんぶ','zenbu','名词')],
'l026':[('全部','ぜんぶ','zenbu','名词'),('を','を','o','助词'),('消したら','けしたら','keshitara','动词（たら形）'),('自由に','じゆうに','jiyuu ni','形容动词'),('なれる','なれる','nareru','动词')],
'l027':[('煩い','うるさい','urusai','形容词'),('脳内','のうない','nounai','名词'),('誰か','だれか','dareka','疑问代词'),('の','の','no','助词'),('言葉','ことば','kotoba','名词')],
'l028':[('捨てたら','すてたら','sutetara','动词（たら形）'),('自分','じぶん','jibun','名词'),('を','を','o','助词'),('生きれる','いきれる','ikireru','动词（可能形）'),('のかな','のかな','no kana','终助词表达')],
'l029':[('立ち止まる','たちどまる','tachidomaru','动词'),('事','こと','koto','名词'),('ばかり','ばかり','bakari','助词'),('覚えてしまった','おぼえてしまった','oboete shimatta','固定表达')],
'l030':[('あれでもない','あれでもない','are demo nai','固定表达'),('これでもない','これでもない','kore demo nai','固定表达')],
'l031':[('歪んでしまった','ゆがんでしまった','yugande shimatta','固定表达')],
'l032':[('後悔ばかりで','こうかいばかりで','koukai bakari de','固定表达'),('夜','よる','yoru','名词'),('が','が','ga','助词'),('終わらない','おわらない','owaranai','动词（否定形）'),('や','や','ya','终助词')],
'l033':[('朝焼けには','あさやけには','asayake ni wa','名词＋助词'),('誰か','だれか','dareka','疑问代词'),('いて','いて','ite','动词（て形）'),('心躍る','こころおどる','kokoro odoru','固定表达')],
'l034':[('代償','だいしょう','daishou','名词'),('なんて','なんて','nante','助词')],
'l035':[('困難だって','こんなんだって','konnan datte','固定表达'),('超えていくって','こえていくって','koete ikutte','固定表达')],
'l036':[('決めたのに','きめたのに','kimeta noni','固定表达'),('壊したくなって','こわしたくなって','kowashitaku natte','固定表达')],
'l037':[('想像の中じゃ','そうぞうのなかじゃ','souzou no naka ja','固定表达'),('足りない','たりない','tarinai','动词（否定形）'),('くらいに','くらいに','kurai ni','助词表达')],
'l038':[('君','きみ','kimi','名词'),('と','と','to','助词'),('叶えてみたい','かなえてみたい','kanaete mitai','动词（てみたい形）')],
'l039':[('感傷','かんしょう','kanshou','名词'),('ばっか','ばっか','bakka','助词')],
'l040':[('なんで','なんで','nande','副词'),('心','こころ','kokoro','名词'),('に','に','ni','助词'),('壁','かべ','kabe','名词'),('を','を','o','助词')],
'l041':[('つくってしまうんだろう','つくってしまうんだろう','tsukutte shimaun darou','固定表达')],
'l042':[('誰か','だれか','dareka','疑问代词'),('の','の','no','助词'),('希望に','きぼうに','kibou ni','名词＋助词'),('なれるなら','なれるなら','nareru nara','动词（条件形）')],
'l043':[('誰にも','だれにも','darenimo','疑问代词＋助词'),('言えない','いえない','ienai','动词（否定形）'),('痛み','いたみ','itami','名词'),('引き連れて','ひきつれて','hikitsurete','动词（て形）'),('さ','さ','sa','终助词')],
'l044':[('ただ','ただ','tada','副词'),('独りでも','ひとりでも','hitori demo','副词性表达')],
'l045':[('紛れもない','まぎれもない','magire mo nai','固定表达'),('自分','じぶん','jibun','名词'),('を','を','o','助词'),('信じて','しんじて','shinjite','动词（て形）')],
'l046':[('傷つき','きずつき','kizutsuki','动词'),('堕ちる','おちる','ochiru','动词'),('たび','たび','tabi','名词'),('歌えばいい','うたえばいい','utaeba ii','固定表达')],
}

# Explicit ruby: literal kanji runs only.  Existing kana remains unannotated.
R={
'l006':[('誰','だれ'),('染','そ'),('誰','だれ')], 'l007':[('夢','ゆめ'),('思','おも'),('笑','わら')], 'l008':[('忘','わす'),('弱','よわ'),('悴','かじか')], 'l009':[('寂','さび'),('言','い')], 'l010':[('会話','かいわ'),('傷','きず'),('心','こころ')], 'l011':[('誰一人','だれひとり'),('誰一人','だれひとり'),('幸','しあわ')], 'l012':[('期待','きたい'),('膨','ふく')], 'l013':[('違','ちが'),('嘘','うそ'),('全部','ぜんぶ'),('捨','す')], 'l014':[('大逸','おおそ')], 'l015':[('聞','き')], 'l016':[('夢','ゆめ'),('幼','おさな')], 'l017':[('今日','きょう'),('変','か')], 'l019':[('君','きみ'),('歌','うた')], 'l020':[('何度','なんど')], 'l021':[('感情','かんじょう'),('揺','ゆ'),('動','うご')], 'l022':[('嫌','きら')], 'l023':[('歩','ある'),('行','い')], 'l024':[('誰','だれ'),('私','わたし'),('胸','むね'),('張','は')], 'l025':[('要','い'),('全部','ぜんぶ')], 'l026':[('全部','ぜんぶ'),('消','け'),('自由','じゆう')], 'l027':[('煩','うるさ'),('脳内','のうない'),('誰','だれ'),('言葉','ことば')], 'l028':[('捨','す'),('自分','じぶん'),('生','い')], 'l029':[('立','た'),('止','ど'),('事','こと'),('覚','おぼ')], 'l031':[('歪','ゆが')], 'l032':[('後悔','こうかい'),('夜','よる'),('終','お')], 'l033':[('朝焼','あさや'),('誰','だれ'),('心躍','こころおど')], 'l034':[('代償','だいしょう')], 'l035':[('困難','こんなん'),('超','こ')], 'l036':[('決','き'),('壊','こわ')], 'l037':[('想像','そうぞう'),('中','なか'),('足','た')], 'l038':[('君','きみ'),('叶','かな')], 'l039':[('感傷','かんしょう')], 'l040':[('心','こころ'),('壁','かべ')], 'l042':[('誰','だれ'),('希望','きぼう')], 'l043':[('誰','だれ'),('言','い'),('痛','いた'),('引','ひ')], 'l044':[('独','ひとり')], 'l045':[('紛','まぎ'),('自分','じぶん'),('信','しん')], 'l046':[('傷','きず'),('堕','お'),('歌','うた')]
}

def rubies(text, pairs):
    result=[]; pos=0
    for base,reading in pairs:
        i=text.find(base,pos)
        if i<0: raise ValueError((text,base))
        result.append({'base':base,'reading':reading,'start':i,'end':i+len(base)})
        pos=i+len(base)
    return result
def card(t,r,ro,pos):
    return {'token':t,'reading':r,'romaji':ro,'grammarStructureZh':pos,'posZh':pos,'render':True,'showJlpt':False,'status':'assisted-proposal','fieldProvenance':{'token':'lexical v2 proposal','reading':'lexical v2 proposal','romaji':'lexical v2 proposal','grammarStructureZh':'lexical v2 proposal'}}
changes=[]
for f in frames:
    fid=f['id']; text=f['caption']['japanese']; units=D[fid]
    expected=text.replace(' ','').replace('？','').replace('?','')
    if ''.join(x[0] for x in units)!=expected:
        raise ValueError(f'{fid}: '+''.join(x[0] for x in units)+' != '+text)
    newcards=[card(*x) for x in units]
    changes.append({'frameId':fid,'field':'grammarCards','old':f['grammarCards'],'new':newcards,'confidence':0.93,'evidence':['JLPT Dictionary / Weblio lexical segmentation cross-check','learnable fixed-expression grouping; exact lyric-character coverage asserted'],'changesTokenStructure':True})
    newruby=rubies(text,R.get(fid,[]))
    changes.append({'frameId':fid,'field':'caption.furigana','old':f['caption'].get('furigana',[]),'new':newruby,'confidence':0.95,'evidence':['kanji-only ruby; literal kana deliberately unannotated'],'changesTokenStructure':False})
    newroma=' '.join(x[2] for x in units)
    changes.append({'frameId':fid,'field':'caption.romaji','old':f['caption'].get('romaji',''),'new':newroma,'confidence':0.95,'evidence':['revised Hepburn romaji aligned to lexical card units'],'changesTokenStructure':False})
out={'schemaVersion':2,'reviewRole':'lexical','scope':{'frameIds':[f['id'] for f in frames]},'baseFrameSha256':sha,'changes':changes}
dst=ROOT/'project'/'proposals'/'lexical.json'; dst.parent.mkdir(exist_ok=True)
dst.write_text(json.dumps(out,ensure_ascii=False,indent=2)+'\n',encoding='utf-8')
print(json.dumps({'output':str(dst),'frames':len(frames),'changes':len(changes),'cards':sum(len(D[f['id']]) for f in frames),'sha':sha},ensure_ascii=False))

File: projects/kanjou-glass/project/work/build-lexical-proposal.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const projectRoot = path.resolve(__dirname, '..', '..');
const framesPath = path.join(projectRoot, 'project', 'frames.json');
const outputPath = path.join(projectRoot, 'project', 'proposals', 'lexical.json');
const expectedHash = '3e3a32c058e5eca15cce097f0ab61965511d3dbf66127431d55ce51e4064af9f';
const raw = fs.readFileSync(framesPath);
const actualHash = crypto.createHash('sha256').update(raw).digest('hex');
if (actualHash !== expectedHash) {
  throw new Error(`frames.json hash changed: expected ${expectedHash}, got ${actualHash}`);
}

const doc = JSON.parse(raw.toString('utf8'));
const byId = new Map(doc.frames.map((frame) => [frame.id, frame]));
const changes = [];

const qrcEvidence = [
  'project/timing/qm.json：冻结日文歌词、字符位置和逐词时间。',
  'project/timing/roma.json：冻结 QQ 音乐唱读轨；本提案以唱读轨而非词典默认音纠正歌曲实际读音。'
];
const rubyEvidence = [
  'japanese-song-study-video 固定前景契约：振假名只覆盖连续汉字段，不重复显示歌词中已有的平假名；纯假名词不显示振假名。',
  'project/timing/qm.json 与 project/timing/roma.json：逐字核对汉字位置和对应唱读。'
];

function add(frameId, field, oldValue, newValue, confidence, evidence, changesTokenStructure = false) {
  if (JSON.stringify(oldValue) === JSON.stringify(newValue)) return;
  changes.push({
    frameId,
    field,
    old: oldValue,
    new: newValue,
    confidence,
    evidence,
    changesTokenStructure
  });
}

function cardField(frameId, index, key, newValue, confidence, evidence) {
  const frame = byId.get(frameId);
  add(frameId, `grammarCards[${index}].${key}`, frame.grammarCards[index][key], newValue, confidence, evidence, false);
}

function captionField(frameId, key, newValue, confidence, evidence) {
  const frame = byId.get(frameId);
  add(frameId, `caption.${key}`, frame.caption[key], newValue, confidence, evidence, false);
}

function ruby(cardIndex, base, reading, tokenStart, baseStart) {
  return {
    cardIndex,
    base,
    surfaceOffset: baseStart - tokenStart,
    reading,
    tokenStart,
    baseStart
  };
}

function furigana(frameId, entries, confidence = 0.99) {
  captionField(frameId, 'furigana', entries, confidence, rubyEvidence);
}

// QRC singing overrides dictionary-default readings.
const minamoEvidence = [
  ...qrcEvidence,
  'roma.json 对本句逐拍记录为「mi na mo」；因此「水面」按歌曲实际唱读标为「みなも / minamo」，不是词典默认的「すいめん」。',
  '小学馆《デジタル大辞泉》收录「みなも【水面】」：https://kotobank.jp/word/%E6%B0%B4%E9%9D%A2-639224'
];
cardField('l001', 0, 'reading', 'みなも', 1.0, minamoEvidence);
cardField('l001', 0, 'romaji', 'minamo', 1.0, minamoEvidence);
captionField('l001', 'romaji', 'minamo no hamon kara kao sorashita no wa', 1.0, minamoEvidence);
furigana('l001', [
  ruby(0, '水面', 'みなも', 0, 0),
  ruby(2, '波紋', 'はもん', 3, 3),
  ruby(4, '顔', 'かお', 8, 8),
  ruby(5, '逸', 'そ', 9, 9)
], 1.0);

const watashiEvidence = [
  ...qrcEvidence,
  'roma.json 对本句逐拍记录为「wa ta shi」；草稿的「わたくし / watakushi」与冻结唱读轨不符。',
  '小学馆《デジタル大辞泉》列出「私」常用读音「わたし」：https://kotobank.jp/word/%E7%A7%81-664668'
];
cardField('l002', 0, 'reading', 'わたし', 1.0, watashiEvidence);
cardField('l002', 0, 'romaji', 'watashi', 1.0, watashiEvidence);
captionField('l002', 'romaji', 'watashi no kokoro yuraida no kakushitakute', 1.0, watashiEvidence);
furigana('l002', [
  ruby(0, '私', 'わたし', 0, 0),
  ruby(2, '心', 'こころ', 2, 2),
  ruby(3, '揺', 'ゆ', 4, 4),
  ruby(5, '隠', 'かく', 9, 9)
], 1.0);

// Kanji-only furigana: do not repeat literal kana already present in the lyric.
furigana('l003', [
  ruby(2, '並', 'なら', 4, 4),
  ruby(5, '心臓', 'しんぞう', 11, 11)
]);

furigana('l007', [
  ruby(0, '袖', 'そで', 0, 0),
  ruby(1, '触', 'ふ', 1, 1),
  ruby(2, '満', 'み', 5, 5),
  ruby(3, '引', 'ひ', 7, 7),
  ruby(4, '目', 'め', 11, 11),
  ruby(6, '合', 'あ', 13, 13)
]);

furigana('l009', [
  ruby(0, '一口', 'ひとくち', 0, 0),
  ruby(1, '味', 'あじ', 2, 2),
  ruby(3, '名前', 'なまえ', 9, 9),
  ruby(5, '呼', 'よ', 12, 12)
]);

furigana('l010', [
  ruby(0, '喉', 'のど', 0, 0),
  ruby(1, '通', 'とお', 1, 1)
]);

furigana('l011', [
  ruby(0, '初', 'はじ', 0, 0),
  ruby(1, '知', 'し', 3, 3),
  ruby(2, '感情', 'かんじょう', 5, 5)
]);

furigana('l013', [
  ruby(0, '溶', 'と', 0, 0),
  ruby(0, '込', 'こ', 0, 2),
  ruby(1, '光', 'ひか', 6, 6)
]);

furigana('l014', [
  ruby(2, '透', 'す', 4, 4)
]);

furigana('l015', [
  ruby(0, '向', 'む', 0, 0),
  ruby(0, '側', 'がわ', 0, 3),
  ruby(1, '覗', 'のぞ', 4, 4)
]);

furigana('l017', [
  ruby(0, '世界', 'せかい', 0, 0),
  ruby(2, '変', 'か', 3, 3),
  ruby(4, '綺麗', 'きれい', 8, 8),
  ruby(7, '思', 'おも', 12, 12)
]);

furigana('l019', [
  ruby(0, '隣', 'とな', 0, 0),
  ruby(0, '合', 'あ', 0, 2),
  ruby(1, '影', 'かげ', 5, 5),
  ruby(3, '影', 'かげ', 7, 7),
  ruby(5, '期待', 'きたい', 12, 12)
]);

// 待ちあぐね is the continuative surface of the single dictionary verb 待ちあぐねる.
{
  const frame = byId.get('l020');
  const oldCards = frame.grammarCards;
  const merged = structuredClone(oldCards[1]);
  merged.token = '待ちあぐね';
  merged.reading = 'まちあぐね';
  merged.romaji = 'machiagune';
  merged.grammarStructureZh = '动词（活用）';
  merged.dictionaryForm = '待ちあぐねる';
  merged.status = 'assisted-proposal';
  merged.reviewRequired = true;
  merged.fieldProvenance = {
    ...merged.fieldProvenance,
    segmentation: 'lexical role proposal: dictionary verb 待ちあぐねる',
    reading: 'project/timing/roma.json + lexical review',
    romaji: 'lexical review'
  };
  const newCards = [oldCards[0], merged, ...oldCards.slice(3)];
  const evidence = [
    ...qrcEvidence,
    '小学馆《デジタル大辞泉》将「待ちあぐねる」收为一个动词词条；歌词「待ちあぐね」是其连用形：https://kotobank.jp/word/%E5%BE%85%E3%81%A1%E5%80%A6%E3%81%AD%E3%82%8B-634561',
    '将「待ち」与「あぐね」分成两个学习词卡会切断词典词形，并使罗马音产生不自然的词内空格。'
  ];
  add('l020', 'grammarCards', oldCards, newCards, 0.99, evidence, true);
  captionField('l020', 'romaji', 'uchikeshite machiagune te o totta', 0.99, evidence);
  furigana('l020', [
    ruby(0, '打', 'う', 0, 0),
    ruby(0, '消', 'け', 0, 2),
    ruby(1, '待', 'ま', 6, 6),
    ruby(2, '手', 'て', 12, 12),
    ruby(4, '取', 'と', 14, 14)
  ]);
}

furigana('l021', [
  ruby(0, '苦', 'にが', 0, 2),
  ruby(2, '甘', 'あま', 7, 7)
]);

furigana('l022', [
  ruby(3, '想', 'おも', 6, 6),
  ruby(5, '胸', 'むね', 9, 9),
  ruby(7, '満', 'み', 11, 11)
]);

furigana('l024', [
  ruby(0, '熱', 'ねつ', 0, 0),
  ruby(2, '持', 'も', 2, 2),
  ruby(5, '惑', 'まど', 9, 9)
]);

furigana('l026', [
  ruby(0, '飛', 'と', 0, 0),
  ruby(0, '越', 'こ', 0, 2)
]);

furigana('l028', [
  ruby(0, '思', 'おも', 0, 0),
  ruby(2, '遠', 'とお', 5, 5)
]);

furigana('l029', [
  ruby(2, '涙', 'なみだ', 4, 4),
  ruby(3, '理由', 'りゆう', 6, 6),
  ruby(5, '独', 'ひと', 10, 10),
  ruby(5, '占', 'じ', 10, 12)
]);

// Repeated lyric bundles must carry identical readings and ruby policy.
furigana('l030', [
  ruby(0, '喉', 'のど', 0, 0),
  ruby(1, '通', 'とお', 1, 1)
]);

furigana('l031', [
  ruby(0, '初', 'はじ', 0, 0),
  ruby(1, '知', 'し', 3, 3),
  ruby(2, '感情', 'かんじょう', 5, 5)
]);

furigana('l033', [
  ruby(0, '溶', 'と', 0, 0),
  ruby(0, '込', 'こ', 0, 2),
  ruby(1, '光', 'ひか', 6, 6)
]);

const proposal = {
  schemaVersion: 2,
  reviewRole: 'lexical',
  scope: {
    frameIds: doc.frames.map((frame) => frame.id)
  },
  baseFrameSha256: expectedHash,
  changes
};

if (proposal.scope.frameIds.length !== 33 || proposal.scope.frameIds[0] !== 'l001' || proposal.scope.frameIds[32] !== 'l033') {
  throw new Error('Expected complete l001-l033 scope.');
}

fs.mkdirSync(path.dirname(outputPath), { recursive: true });
fs.writeFileSync(outputPath, `${JSON.stringify(proposal, null, 2)}\n`, { encoding: 'utf8' });
console.log(JSON.stringify({ outputPath, changeCount: changes.length, scopeCount: proposal.scope.frameIds.length }, null, 2));
